### PR TITLE
Extend Cardano.Api.ProtocolParameters with Alonzo params

### PR DESCRIPTION
**Pull request by dcoutts**
* Add PlutusScriptV1 constructor in the existing `PlutusScriptVersion` type which previously had no constructors at all.
* Add API types for execution units, prices and cost models
* Extend the ProtocolParameters with Alonzo era ones, both `ProtocolParameters` and `ProtocolParametersUpdate`.

### CODE REVIEW

**Comment by dcoutts** [comment]
Note that this extends the branch for PR #2657 so ignore the first 3 commits, and this will need to be rebased after that goes in.

This PR is a rebased and tidied up version of PR #2634 by @Jimbo4350. If we're happy with this one we can close the original. (I opened it as a new one entirely since it's actually substantially different so I didn't want to just force push over the top.)

File: cardano-api/src/Cardano/Api/ProtocolParameters.hs
Comment (jonathanknowles) [inline cardano-api/src/Cardano/Api/ProtocolParameters.hs:238]: Hi @dcoutts

Does "word" in this context refer to the size of a [`Data.Word`](http://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Word.html#t:Word)?

(I notice that we still use "byte" elsewhere, for example, in `txFeePerByte`, so I'm guessing the differentiation is deliberate in some way.)

Apologies for the drive-by comment, the PR title piqued my interest and decided to have a look. 😄

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs
Comment (Jimbo4350) [inline cardano-node/src/Cardano/Node/Protocol/Cardano.hs:182]: Is this to enable testing of a hardfork to Alonzo before consensus has merged their changes?

**Review by cblp — COMMENTED** [review]
LGTM, but somebody else should approve

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs
Comment (dcoutts) [reply cardano-node/src/Cardano/Node/Protocol/Cardano.hs]: Yes. Though I've now moved this patch into this PR instead: https://github.com/input-output-hk/cardano-node/pull/2400

**Comment by dcoutts** [comment]
Rebased on top of #2649 since this order makes more sense. Only the last three commits are part of this PR proper. All the previous commits are from #2657 and #2649 that this builds on.

File: cardano-api/src/Cardano/Api/ProtocolParameters.hs
Comment (newhoggy) [reply cardano-api/src/Cardano/Api/ProtocolParameters.hs:238]: > In Alonzo, the protocol parameter `minUTxOValue` is deprecated, and replaced by `adaPerUTxOWord`.
This specifies directly the deposit required for storing bytes of data on the ledger in the form of
UTxO entries

It's not clear to me that word means byte, but _I think so_.

File: cardano-cli/src/Cardano/CLI/Shelley/Script.hs
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Script.hs:33]: Typo `Syncax`

**Review by newhoggy — APPROVED** [review]
LGTM

**Comment by newhoggy** [comment]
bors merge

**Pull request merged**
